### ISSUE TITLE
RFC: microservice / plugin backend

### ISSUE THREAD

**Issue by anbraten** [kiel-live/kiel-live#31]
# General idea
By the wish to add more data sources to the app the idea to refactor the backend to use some kind of microservice architecture came up. 
In this concept the backend would be serving generic data and data-changes from different providers (like KVG, DB) while being provider independent. Provider collectors could be written in any language being somehow able to send updates via websocket.

## Implementation ideas

The backend should be as simple and stupid as possible by using a super generic web-interface (socket). This way it could be easily extendable without the need to re-deploy the main backend each time a new provider is added or needs updates.

For this a pub-sub mechanism could be used for all app-clients. A client should be able to connect to the backend, subscribe a channel and receive an initial data set which could be stored in some in-memory database included in the server or by using something like Redis. If the provider service now publishes changes to the channel, the backend distributes these changes to all subscribed clients.

### Stop

A `stop` is a fixed point their for example a bus stop or a car-sharing parking spot is located.

```json5
{
  "id": "<id:string>",
  "name": "<name:string>",
  "provider": "<provider:string>",
  "type": "<type:string>", // BUS_STOP, PARKING_SPOT, FERRY_STOP, TRAIN_STOP, SUBWAY_STOP
  "routes": [
    "<route-id:string>" // list of routes using this stop
  ],
  "alerts": [
    "<alert:string>", // general alerts for this stop
  ],
  "arrivals": [
    {
      "name": "<name:string>",
      "vehicle": "<vehicle-id:string>",
      "trip": "<trip-id:string>",
      "route": "<route-id:string>",
      "direction": "<direction:string>",
      "state": "<state:string>", // PREDICTED, STOPPING, DEPARTED
      "planned": "<planned:date>",
      "eta?": "<eta:number(seconds)>"
    }
  ],
  "location": {
    "longitude": "<longitude:number(54.306)>",
    "latitude": "<latitude:number(10.149)>",
  },
  "extra": "<*:*>"
}
```

### Vehicle

A `vehicle` can be of a specific type (exp. bus, bike).

```json5
{
  "id": "<id:string>",
  "name": "<name:string>",
  "type": "<type:string>", // BUS, BIKE, CAR, ESCOOTER, FERRY, TRAIN, SUBWAY
  "active": "<active:boolean>",
  "state?": "<state:string>",
  "battery?": "<battery:number(percent)>",
  "location": {
    "heading": "<heading:number(degree)>",
    "longitude": "<longitude:number(54.306)>",
    "latitude": "<latitude:number(10.149)>",
  },
  "trip?": "<trip-id:string>",
  "extra": "<*:*>"
}
```

### Trip

A `trip` is a tour represented by a list of `stops` executed by a `vehicle` (exp. bus) on a specific `route`.

```json5
{
  "id": "<id:string>",
  "vehicle": "<vehicle-id:string>",
  "direction": "<direction:string>",
  "stops": [
    {
      "id": "<stop-id:string>",
      "name": "<name:string>",
      "state": "<state:string>", // PREDICTED, STOPPING, DEPARTED
      "planned": "<planned:date>",
      "eta?": "<eta:number(seconds)>" // needed?
    }
  ],
  "extra": "<*:*>"
}
```

### Route

A `route` is a fixed list of waypoints a vehicle could drive along. A one-time tour on a specific `route` is called a `trip`.

```json5
{
  "id": "<id:string>",
  "name": "<name:string>",
  "type": "<type:string>", // BUS, FERRY, TRAIN, SUBWAY
  "active": "<active:boolean>",
  "stops": [
    {
      "id": "<stop-id:string>",
      "location": {
        "longitude": "<longitude:number(54.306)>",
        "latitude": "<latitude:number(10.149)>",
      }
    }
  ],
  "extra": "<*:*>"
}
```

## Possible channels

- `/stops/<provider>`: List of all stops with a `name`, `provider`, `id`,  `type` (bus-stop, car-sharing-station, ferry-pier), `longitude`, `latitude` 
- `/vehicles/<provider>`: List of all vehicles with  a `name`, `provider`, `id`,  `type` (bus, ferry, e-scooter), `longitude`, `latitude` 
- `/stop/<provider>/<stop-id>`: Details of a specific stop
- `/vehicle/<provider>/<vehicle-id>`: Details of a specific vehicle
- `/trip/<provider>/<trip-id>`: Details of a specific trip
- `/route/<provider>/<route-id>`: Details of a specific route
- `/subscribed-channels`: Special channel with a list of all currently subscribed channels (see problems why this could be helpful)

## Possibly interesting providers
For example it would be nice to include e-scooters, bike-sharing and car-sharing. Currently this would be the following providers:
- KVG buses (as already used by the app)
- Tier (e-scooter) [API](https://www.notion.so/TIER-API-Version-1-0a9bfc571e8f41859d994e736a51315b#f114ed57abed4cdbb7abd40dd33a0fc2) - [Docs](https://github.com/ubahnverleih/WoBike/blob/master/Tier.md)
- Sprottenflotte / Nextbike (bike-sharing) [API](https://maps.nextbike.net/maps/nextbike-live.json?city=613&list_cities=0) - [Docs](https://github.com/ubahnverleih/WoBike/blob/master/Nextbike.md)
- Stattauto (car-sharing) [Map /  Booking-system](https://ewi3-stattauto-kiel-luebeck.cantamen.de/)
- Ferries in Kiel by using AIS data

## Problems & limitations
- KVG has some very specific endpoints which need ids from something like the stop being queried. As querying 100 stops at a refresh interval of 3 seconds, we will easily reach rate-limits and risk ip-bans.
  -  __Possible solution__: On demand queries for details requests like stop details. If the stop is currently viewed by some client, the client would subscribe to the stop channel with a name like `/stop/kvg/id123`. The backend could manage a channel with a list of all subscribed channels, so the KVG provider could listen to this special channel and test by some regex for all `/stop/kvg/*` channels and load their needed data.

**Comment by anbraten** [comment]
@lukashass @KalleJoP I created this RFC for a new backend era. 😉  Please feel free to suggest ideas, point out possible problems. 

The current backend is implemented / hacked together as simple Node.js Javascript application, but I think this would be the perfect chance to consider other languages as well.
My suggestions for now would be one of the following:
- Go
- Elixir (phoenixframework)
- Node.js Typescript

What do you think?

**Comment by anbraten** [comment]
solved by #127
